File: Stack/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:

        def opposite_of(chr):
            if chr == ')':
                return '('
            elif chr == ']':
                return '['
            else:
                return '{'
         
        stack = []
        
        if len(s)==1 or len(s)%2 != 0:
            return False
        

        for i in range(len(s) ):
            # if next is open add it
            if s[i] in ('(', '[', '{'):
                stack.append(s[i])
            # if next is closing, then it should be last opened

            else:
                if len(stack)==0 or stack.pop() != opposite_of( s[i] ) :
                    # print('returing false')
                    return False

            # print(f'{s[i]} := {stack}')


        if len(stack)== 0:
            return True
        else:
            return False

File: Stack/valid-parentheses.java
class Solution {
    public char opp_of(char c){
        if (c==')') 
            return '(';
        else if (c=='}')
            return '{';

        else return '[';
    }

    public boolean isValid(String s) {
        // make a char array of the string

        char[] s_chr = s.toCharArray();
        char[] stack = new char[s_chr.length];
        int stack_i = 0;

        for(int i= 0; i< s_chr.length; i++){
            if (s_chr[i] == '(' || s_chr[i] == '{' || s_chr[i] == '['){
                stack[stack_i ++] = s_chr[i];
                 
            }
            else{
                if (stack_i == 0 ||  stack[stack_i-1] != opp_of(s_chr[i])){
                    return false;
                }
                else{
                    stack_i--;
                }
            }
        }

        if (stack_i == 0)
            return true;
        else 
            return false;

    }
}